### PR TITLE
Coupon Codes API Project

### LINKED ISSUE #151: Create M2 Final Project
Builds off the Little Shop group project (which has a BE and FE).  Final will be mostly Rails practice with a one to many relationship plus a small amount of FE

BE: (Erin)
Coupon table with merchant id; endpoints to crud coupons; some db operations (ex: discounted prices)

Front End: (Kayla)
Group project starting point: 
Students will continue to build off the same FE they used for the Group project.  In the group project,  the FE only has CRUD functionality for the merchants.  Merchants are displayed, can be edited, deleted, and new merchants can be added.   Users can view all the items for a specific merchant.

Final Project FE requirements to add onto what they have from the group project:
Students should build out the following:
- Users can click on a merchant (or a button for a given merchant) and see all the coupons for that merchant.  No edit/add/delete functionality needed.  
- We could provide a comp (or a few comp options) for the display of the merchant coupons might look like.  If design isn't their strong suit, we don't want those visual design decisions jamming them up when that isn't a learning goal.
- We should provide the code for the fetch GET request, including that data in the Promise.All, adding it to a global variable.  Students will simply access data through the global variable and display it on the DOM.
- Rather than having this already built into the boilerplate, the spec could walk them through copy.pasting the relevant code into their FE application's existing code.  This forced them to give attention to the network requests without having the challenge of putting it together themself.  As part of their eval, they could be required to describe what is happening in the provided code they copy/pasted into their application.

### PR DESCRIPTION
**Pull request by epintozzi**
# Curriculum PR Template

### Does this PR close any issues?
*This PR closes #151 *

### Description
Adapts existing Coupon Codes project to an API BE with separate FE component

### What questions do you have/what do you want feedback on? (optional)
@juliet-e Thoughts on the rubric? I copied from the group project and tweaked verbiage for solo project presentation. Will these be live presentations?

Feedback on typos, readability, and the like

Remaining todos:
* Link group project starter repo
* Add FE component and update rubric accordingly

### CODE REVIEW

File: module2/projects/coupon-codes-api/evaluation.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+With remaining time, instructors will ask questions and follow-up on feedback requests.
+
+
+## Rubric
+Your project will be evaluated based on the following rubric: 
+
+### Feature Delivery
+
+* Exceeds Expectations: Project completes all requirements and at least one extensions.
```
Comment (juliet-e) [inline module2/projects/coupon-codes-api/evaluation.md:30]: There's just an extra `s` at the end of this sentence (extension vs extensions)

File: module2/projects/coupon-codes-api/evaluation.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+* Meets Expectations: Project completes all requirements. Postman tests pass, and FE functions correctly when wired up to API. 
+* Approaching Expectations: Project fails to complete 1 - 2 required endpoints
+* Below Expectations: Project fails to complete more than 3 or more endpoints
+
+### Project Management
+
+* Exceeds Expectations: GitHub Project Board is fully up to date in all checkins and the evaluation. Live presentation is prepared with visuals, a logical flow, and distributed talking points. Students complete all requirements in "Meets Expectations" as well
+* Meets Expectations: Students use GitHub projects or similar tool to track all project tasks. Project board is mostly up to date in all checkins, and fully up to date in evaluation. Students participate in meaningful code review and it's visible through PR comments or live meetings (students can speak to one of these meetings). Live presentation is prepared, and follows a logical flow. All group members participate, but maybe with some instructor prompting.
```
Comment (juliet-e) [inline module2/projects/coupon-codes-api/evaluation.md:38]: Do you think we should evaluate students on project management for a solo project? Do we currently do that? I think evaluation of some planning would be good but not sure what makes the most sense for solo.

Comment (juliet-e) [reply module2/projects/coupon-codes-api/evaluation.md:38]: Looks like we currently evaluation on technical presentation instead of project management here -> https://backend.turing.edu/module2/projects/coupon_codes/evaluation

I think I'm I'd learn more towards technical presentation rather than proj management for this project

File: module2/projects/coupon-codes-api/index.md [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+* Additional gems to be added to the project must have instructor approval.
+  * RSpec, Pry, Simplecov, Shoulda-Matchers, Orderly, jsonapi-serializer (or other serializer gem), HTTParty, Faker, Faraday, and FactoryBot are pre-approved
+* Scaffolding is not permitted on this project.
+
+## Setup
+
+This project is an extension of Little Shop. Students have two options for setup:
+
+1. If your Little Shop project is complete, you can use it as a starting point for this project. If you are not the repo owner, fork the project to your account. If you are the repo owner, you can work off the repo without forking, just make sure your teammates have a chance to fork before pushing any commits to your repo.
+1. If your Little Shop project is _not_ complete, fork ****Update** [this repo](need to update)** as a starting point for this project.
```
Comment (juliet-e) [inline module2/projects/coupon-codes-api/index.md:29]: Here's the starter! https://github.com/turingschool-examples/little-shop-be

File: module2/projects/coupon-codes-api/index.md [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+
+-----
+
+## Functionality Overview
+
+* A Coupon belongs to a Merchant
+* An Invoice optionally belongs to a Coupon. An invoice may only have one coupon.
+  * Note: When creating this new association on Invoice, your existing tests will fail unless the association is optional. Use these guides as a reference.
+  * You are not required to build functionality for a user applying a coupon to an invoice, but can instead use test data, Rails console or seed data to add coupons to existing invoices to verify behavior.
+* You should be full CRUD functionality for coupons with criteria/restrictions defined below:
```
Comment (juliet-e) [inline module2/projects/coupon-codes-api/index.md:42]: Should `be` instead be `build`?

File: module2/projects/coupon-codes-api/index.md [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+
+{
+  "data": {
+    "id": "1",
+    "type": "coupon",
+    "attributes": {
+      "name": "Buy One Get One 50",
+      "code": "BOGO50",
+      "amount_off": 50,
+      "percent_off": null,
```
Comment (juliet-e) [inline module2/projects/coupon-codes-api/index.md:80]: I'd be curious what @abdulredd thinks of this schema cause he's run this existing coupon code project more than me. I've seen students use a `type` attribute (percent or dollar) and then an `amount` attribute which is the integer amount. So `type` of percent and `amount` of 20 would be 20% off, for example. This would avoid nullable fields. I think, by convention, if a field is null, it's also just not included in a JSON response. So we could try that out if we keep the schema like this. 

Part of the interesting part of the project is students working through how to design this in the database, so I'm wondering if we should have some flexibility in how the attributes are displayed? I guess we can't if we require it to work on the FE though...

File: module2/projects/coupon-codes-api/index.md [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+      "status": "inactive",
+      "merchant_id": 1,
+      "uses": 1
+    }
+  }
+}
+```
+
+<!-- **Sad Paths to consider:**
+1. A coupon cannot be deactivated if there are any pending invoices with that coupon. -->
```
Comment (juliet-e) [inline module2/projects/coupon-codes-api/index.md:241]: I think this is a good idea for students to consider! Is  it commented out because you took it out?

File: module2/projects/coupon-codes-api/index.md [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+  ]
+}
+```
+
+</section>
+
+<section class="dropdown">
+### 8. Merchants
+
+Update the merchants endpoint to include a count of coupons for each merchant and a count of invoices with coupons applied for each merchant.
```
Comment (juliet-e) [inline module2/projects/coupon-codes-api/index.md:402]: Love these endpoints that require extending existing endpoints!

File: module2/projects/coupon-codes-api/index.md (hunk 5 above)
Comment (epintozzi) [reply module2/projects/coupon-codes-api/index.md:80]: Yeah I also would probably do a type and then amount. But I didn't know if that felt too complicated, but it's probably not. Like you said, it gives them a chance to make and talk about design decisions. Perhaps one of the FE tasks can be adjusting the FE code to add and/or properly center the coupon value?

File: module2/projects/coupon-codes-api/index.md (hunk 6 above)
Comment (epintozzi) [reply module2/projects/coupon-codes-api/index.md:241]: Yes I commented it out because I was grappling with whether the relationships were ones the students could handle since I am not the most familiar with the db structure of this project. If you think it's good, I'll add it back in!

File: module2/projects/coupon-codes-api/index.md (hunk 3 above)
Comment (epintozzi) [reply module2/projects/coupon-codes-api/index.md:29]: Thanks!

File: module2/projects/coupon-codes-api/evaluation.md (hunk 2 above)
Comment (epintozzi) [reply module2/projects/coupon-codes-api/evaluation.md:38]: Good idea that makes a lot of sense. I largely copied over the rubric from the group project to get this up for feedback. I will update to evaluate the presentation instead. FWIW I think it's okay and good for students to practice PM on solo projects but given the timeline for this one, it may end up just feeling like overhead for them.

File: module2/projects/coupon-codes-api/evaluation.md (hunk 1 above)
Comment (epintozzi) [reply module2/projects/coupon-codes-api/evaluation.md:30]: Thank you. Typos are my specialty.

File: module2/projects/coupon-codes-api/index.md (hunk 5 above)
Comment (epintozzi) [reply module2/projects/coupon-codes-api/index.md:80]: Here's what I'm thinking for this after a little more time mulling it over.

The FE portion of this project is students adding coupon info to their existing merchant pages (or whatever makes sense with what they were given for the group project). This way, they also have to design the schema for the Coupon table which is an important skill for us to assess.

We can either remove example responses altogether, or I can edit them to look something like this:
```bash
status: 200
body:

{
  "data": {
    "id": "1",
    "type": "coupon",
    "attributes": {
      "name": "Buy One Get One 50",
      ...remaining attributes go here...
    }
  }
}
```

Thoughts?

File: module2/projects/coupon-codes-api/index.md (hunk 7 above)
Comment (abdulredd) [reply module2/projects/coupon-codes-api/index.md:402]: Same!

Quick question, for task 7 ad maybe even 8, have we removed the requirements for showing the subtotal (total w/o coupons) and grand total (total w/ coupons)?

File: module2/projects/coupon-codes-api/evaluation.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+## Rubric
+Your project will be evaluated based on the following rubric: 
+
+### Feature Delivery
+
+* Exceeds Expectations: Project completes all requirements and at least one extension.
+* Meets Expectations: Project completes all requirements. Postman tests pass, and FE functions correctly when wired up to API. 
+* Approaching Expectations: Project fails to complete 1 - 2 required endpoints
+* Below Expectations: Project fails to complete more than 3 or more endpoints
```
Comment (abdulredd) [inline module2/projects/coupon-codes-api/evaluation.md:33]: Suggestion: 
* Below Expectations: Project fails to complete 3 or more endpoints

File: module2/projects/coupon-codes-api/evaluation.md (hunk 2 above)
Comment (juliet-e) [reply module2/projects/coupon-codes-api/evaluation.md]: Totally -- agreed!

File: module2/projects/coupon-codes-api/index.md (hunk 6 above)
Comment (juliet-e) [reply module2/projects/coupon-codes-api/index.md]: I think that might be a bit tricky but definitely doable for them! This is still just multiple one-to-many relationship style joins.

File: module2/projects/coupon-codes-api/evaluation.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+* Exceeds Expectations: Project achieves 100% test coverage at the unit and integration levels. Project utilizes new testing tool such as Faker or FactoryBot.
+* Meets Expectations: Project achieves greater than 98% test coverage. The tests include sad path and edge case testing.
+* Approaching Expectations: Project achieves greater than 90% test coverage
+* Below Expectations: Project does not have 90% test coverage.
+
+### Technical Quality
+
+* Exceeds Expectations: Project demonstrates exceptionally well factored code, utilizing strategies or tools not discussed in class. 
+* Meets Expectations: Project demonstrates solid code quality, MVC principles (1-2 infractions is okay), and uses a serializer for formatting JSON responses. RESTful endpoints follow RESTful routing and design. Error handling is refactored to use `rescue_from` blocks and helper methods where appropriate
```
[the remark this answers is not in the document]
Comment (juliet-e) [reply module2/projects/coupon-codes-api/evaluation.md:52]: IMO we can give them the path and evaluate that they are building their routes and controllers RESTfully. REST will not be as challenging for them without many-to-many and the redirects that happen in a monolith, but I think that's okay?

File: module2/projects/coupon-codes-api/index.md [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+Below is the expected JSON response for each request. We have also outlined a few examples of Sad Paths you may consider adding in. In your project, you should take time to implement at least 2 sad paths total, but you are not limited to the examples we provide. 
+
+<section class="dropdown">
+### 1. Merchant Coupon Show Page 
+
+Returns a specific coupon and shows a count of how many times that coupon has been used.
+
+**Request**
+```bash
+GET /api/v1/merchants/:merchant_id/coupons/:id
```
Comment (juliet-e) [inline module2/projects/coupon-codes-api/index.md:61]: To @abdulredd's comment, I guess we could just have the endpoint description (without the `GET /api/v1/merchants/:merchant_id/coupons/:id` explicitly listed for them) and they have to determine the most RESTful path? I could go either way on this

File: module2/projects/coupon-codes-api/index.md [hunk 11]
```diff
@@ -0,0 +1,477 @@
...
+* Scaffolding is not permitted on this project.
+
+## Setup
+
+This project is an extension of Little Shop. Students have two options for setup:
+
+1. If your Little Shop project is complete, you can use it as a starting point for this project. If you are not the repo owner, fork the project to your account. If you are the repo owner, you can work off the repo without forking, just make sure your teammates have a chance to fork before pushing any commits to your repo.
+1. If your Little Shop project is _not_ complete, fork **[this repo](https://github.com/turingschool-examples/little-shop-be-final-starter)** as a starting point for this project.
+
+## Database Notes
```
Comment (juliet-e) [inline module2/projects/coupon-codes-api/index.md:31]: FYI I added this section because I ran into it yesterday setting up the coupons functionality for Kayla!

File: module2/projects/coupon-codes-api/index.md (hunk 7 above)
Comment (epintozzi) [reply module2/projects/coupon-codes-api/index.md:412]: I think we have because that would involve more complex AR than we'll expect them to know. Does that sound right @juliet-e ?

File: module2/projects/coupon-codes-api/index.md (hunk 10 above)
Comment (epintozzi) [reply module2/projects/coupon-codes-api/index.md:71]: I'll go ahead an remove the request examples for the CRUD endpoints.

**Comment by kaylagordon** [comment]
Changes I just pushed:
- Renamed project to just "Coupon Codes" since it's not just BE now
- Moved BE requirements and FE requirements to separate pages to mimic Little Shop spec
- Added FE requirements and "fetch walkthrough" (@hfaerber  let me know what you think!)
- Updated extensions on main project spec page (@epintozzi  can you confirm that I listed those extensions correctly?)

**Comment by kaylagordon** [comment]
New links:

- `/module2/projects/coupon-codes/index`
- `/module2/projects/coupon-codes/back_end_requirements`
- `/module2/projects/coupon-codes/front_end_requirements`

**Comment by kaylagordon** [comment]
Changes I just pushed:
- Renamed project to just "Coupon Codes" since it's not just BE now
- Moved BE requirements and FE requirements to separate pages to mimic Little Shop spec
- Added FE requirements and "fetch walkthrough" (@hfaerber let me know what you think!)
- Added one like to the evaluations to require students to run FE app in browser (@hfaerber let me know what you think!)
- Updated extensions on main project spec page (@epintozzi  can you confirm that I listed those extensions correctly?)

**Comment by kaylagordon** [comment]
New links:

- `/module2/projects/coupon-codes/index`
- `/module2/projects/coupon-codes/back_end_requirements`
- `/module2/projects/coupon-codes/front_end_requirements`
- `/module2/projects/coupon-codes/evaluation`

**Comment by kaylagordon** [comment]
New boilerplate for FE: https://github.com/turingschool-examples/little-shop-fe-final-starter

**Comment by epintozzi** [comment]
> * Updated extensions on main project spec page (@epintozzi  can you confirm that I listed those extensions correctly?)

Extensions look good to me!
